src/psyclone/psyir/transformations/scalarization_trans.py

```diff
@@ -0,0 +1,451 @@
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: A. B. G. Chalk, STFC Daresbury Lab
+
+'''This module provides the sclarization transformation class.'''
+
+import itertools
+from typing import Optional, Dict, Any, List, Tuple
+
+from psyclone.core import VariablesAccessInfo, Signature, SymbolicMaths
+from psyclone.psyGen import Kern
+from psyclone.psyir.nodes import Call, CodeBlock, Literal, \
+        IfBlock, Loop, Node, Range, Reference, Routine, StructureReference
+from psyclone.psyir.nodes.array_mixin import ArrayMixin
+from psyclone.psyir.symbols import DataSymbol, RoutineSymbol, INTEGER_TYPE
+from psyclone.psyir.transformations.loop_trans import LoopTrans
+
+
+class ScalarizationTrans(LoopTrans):
+    '''This transformation takes a Loop and converts any array accesses
+    to scalar if the results of the loop are unused, and the initial value
+    is unused. For example:
+
+    >>> from psyclone.psyir.backend.fortran import FortranWriter
+    >>> from psyclone.psyir.frontend.fortran import FortranReader
+    >>> from psyclone.psyir.transformations import ScalarizationTrans
+    >>> from psyclone.psyir.nodes import Loop
+    >>> code = """program test
+    ... integer :: i,j
+    ... real :: a(100), b(100)
+    ... do i = 1,100
+    ...   a(i) = i
+    ...   b(i) = a(i) * a(i)
+    ... end do
+    ... do j = 1, 100
+    ...  if(b(i) > 200) then
+    ...    print *, b(i)
+    ...  end if
+    ... end do
```

Comment:
I think I prefer to remove this second loop and the resulting output below. We do not document cases that fail to validate in other transformation examples and it is unclear why this is here in the generated documentation without context.

Reply:
It was to show a case that didn't scalarise, but I can remove the references to b

Reply:
Actually I don't like removing it because it no longer has any context behind why the transformation does anything. If you still want me to I can remove it but I think showing a case where it only scalarizes some elements makes more sense conceptually. I'll add some more documentation explaining it.

Reply:
ah, then it's ok